icicle_v3/backend/cpu/src/curve/cpu_msm.hpp

```diff
@@ -0,0 +1,667 @@
+
+  /**
+   * @brief Setting up phase 2 class members according to phase 1 results.
+   * NOTE: This setup function also potentially blocks until the previous MSM's phase 3 finishes (Given that in batched
+   * MSM, phase 3 is initiated on a separate thread while the next MSM begins phase 1).
+   */
+  void phase2_setup(std::vector<BmSumSegment>& segments);
+
+  /**
+   * @brief Final accumulation required for MSM calculation. the function will either launch a thread to perform the
+   * calculation (`phase3_tread` function) or by the main thread, depending on the position in the batch (Last msm or
+   * not).
+   * @param segments_ptr - Per to vector containing the segment calculations of phase 2
+   * @param idx_in_batch - idx of the current MSM in the batch.
+   * @param result - output, Per to write the MSM result to. Memory for the Per has already been allocated by
+   * the user.
+   */
+  void final_accumulator(std::shared_ptr<std::vector<BmSumSegment>>& segments_ptr, int idx_in_batch, P* result);
+
+  /**
+   * @brief Phase 3 function to be ran by a separate thread (or main in the end of the run) - i.e. without using the
+   * tasks manager. It performs the remaining serial addition in each BM and sums them to one final MSM result.
+   * @param result - Per to write the MSM result to. Memory for the Per has already been allocated by the user.
+   */
+  void phase3_thread(std::shared_ptr<std::vector<BmSumSegment>> segments_ptr, P* result);
+
+  /**
+   * @brief Function for resetting class members between batch runs.
+   */
+  void batch_run_reset() { std::fill(m_bkts_occupancy.begin(), m_bkts_occupancy.end(), false); }
+};
+
+template <typename A, typename P>
+Msm<A, P>::Msm(const MSMConfig& config)
+    : manager(
+        config.ext->get<int>("n_threads") > 0 ? config.ext->get<int>("n_threads")
+                                              : std::thread::hardware_concurrency()),
+
+      m_c(config.c), m_num_bkts(1 << (m_c - 1)), m_precompute_factor(config.precompute_factor),
+      m_num_bms(((scalar_t::NBITS - 1) / (config.precompute_factor * m_c)) + 1),
+      m_are_scalars_mont(config.are_scalars_montgomery_form), m_are_Ps_mont(config.are_points_montgomery_form),
+      m_batch_size(config.batch_size),
+
+      m_buckets(m_num_bms * m_num_bkts), m_bkts_occupancy(m_num_bms * m_num_bkts, false),
+
+      m_log_num_segments(std::max(
+        (int)std::floor(
+          std::log2((double)(config.ext->get<int>("n_threads") * TASKS_PER_THREAD - 1) / (double)(2 * m_num_bms))),
+        0)),
+      m_num_bm_segments(std::min((int)(1 << m_log_num_segments), (int)(m_num_bms * m_num_bkts))),
+      m_segment_size(std::max((int)(m_num_bkts >> m_log_num_segments), 1)),
+
+      m_p3_threads(m_batch_size - 1)
+{
+}
+
+template <typename A, typename P>
+void Msm<A, P>::run_msm(
+  const scalar_t* scalars, const A* bases, const unsigned int msm_size, const unsigned int batch_idx, P* results)
+{
```

Comment:
Those are the phases? please add comment here explaining the top level logic of the algorithm